adapters/yssp/yssp.go

```diff
@@ -0,0 +1,213 @@
+		reqCopy.Imp = []openrtb2.Imp{imp}
+
+		if request.Site != nil {
+			siteCopy := *request.Site
+			reqCopy.Site = &siteCopy
+		} else if request.App != nil {
+			appCopy := *request.App
+			reqCopy.App = &appCopy
+		}
+
+		if err := changeRequestForBidService(&reqCopy, &ysspExt); err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		reqJSON, err := json.Marshal(&reqCopy)
+		if err != nil {
+			errors = append(errors, err)
+			continue
+		}
+
+		reqs = append(reqs, &adapters.RequestData{
+			Method:  "POST",
+			Uri:     a.URI,
+			Body:    reqJSON,
+			Headers: headers,
+		})
+	}
+
+	return reqs, errors
+}
+
+func (a *YSSPAdapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Bad server response: %d.", err),
+		}}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(len(internalRequest.Imp))
+
+	for _, sb := range bidResp.SeatBid {
+		for _, bid := range sb.Bid {
+			exists, mediaTypeId := getImpInfo(bid.ImpID, internalRequest.Imp)
+			if !exists {
+				return nil, []error{&errortypes.BadServerResponse{
+					Message: fmt.Sprintf("Unknown ad unit code '%s'", bid.ImpID),
+				}}
+			}
+
+			if openrtb_ext.BidTypeBanner != mediaTypeId {
+				//only banner is supported, anything else is ignored
+				continue
+			}
+
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &bid,
+				BidType: openrtb_ext.BidTypeBanner,
+			})
+		}
+	}
+
+	return bidResponse, nil
+}
+
+func getImpInfo(impId string, imps []openrtb2.Imp) (bool, openrtb_ext.BidType) {
+	var mediaType openrtb_ext.BidType
+	var exists bool
+	for _, imp := range imps {
+		if imp.ID == impId {
+			exists = true
+			if imp.Banner != nil {
+				mediaType = openrtb_ext.BidTypeBanner
+			}
+			break
+		}
+	}
+	return exists, mediaType
+}
+
+func changeRequestForBidService(request *openrtb2.BidRequest, extension *openrtb_ext.ExtImpYSSP) error {
+	/* Always override the tag ID and (site ID or app ID) of the request */
+	request.Imp[0].TagID = extension.Pos
+	if request.Site != nil {
+		request.Site.ID = extension.Dcn
+	} else if request.App != nil {
+		request.App.ID = extension.Dcn
+	}
+
+	if request.Imp[0].Banner == nil {
+		return nil
+	}
+
+	banner := *request.Imp[0].Banner
+	request.Imp[0].Banner = &banner
+
+	if banner.W != nil && banner.H != nil {
+		if *banner.W == 0 || *banner.H == 0 {
+			return errors.New(fmt.Sprintf("Invalid sizes provided for Banner %dx%d", *banner.W, *banner.H))
```

Comment:
Please use `fmt.Errorf` here instead:
```
return fmt.Errorf("Invalid sizes provided for Banner %dx%d", *banner.W, *banner.H)
```

Reply:
I addressed this, what is the benefit here? semantically we say we have a formatting error ?

Reply:
Correct. It's a better choice to use the most specific method. Since using `errors.New` and `fmt.Sprintf` together is a very common use case, Go has `fmt.Errorf` to make this more explicit.